src/Driver.cpp

```diff
@@ -247,6 +257,73 @@ void Driver::openIPServer(int port, addrinfo const& hints)
     setMainStream(new UDPServerStream(sfd,true));
 }
 
+void find_addr(const char *hostname, const char *port, addrinfo const& hints, struct sockaddr *addr, size_t *addr_len)
+{
+    struct addrinfo *result;
+    int ret = getaddrinfo(hostname, port, &hints, &result);
+    if (ret != 0)
+        throw UnixError("cannot resolve client port " + string(port));
+
+    int sfd = -1;
+    struct addrinfo *rp;
+    for (rp = result; rp != NULL; rp = rp->ai_next) {
+        sfd = socket(rp->ai_family, rp->ai_socktype,
+                rp->ai_protocol);
+        if (sfd == -1)
+            continue;
+
+        if (connect(sfd, rp->ai_addr, rp->ai_addrlen) == 0)
+            break;                  /* Success */
+
+        ::close(sfd);
+    }
+
+    if (rp == NULL)
+    {
+        freeaddrinfo(result);
+        throw UnixError("cannot open client socket on port " + string(port));
+    }
+
+    memcpy(addr, rp->ai_addr, rp->ai_addrlen);
+    *addr_len = rp->ai_addrlen;
+
+    freeaddrinfo(result);
+    close(sfd);
+}
+
+void Driver::openIPBidirectional(std::string const& hostname, int out_port, addrinfo const& out_hints, int in_port, addrinfo const& in_hints)
```

Comment:
This method creates a UDPServerStream ... Probably best would be to move the meat of the method into a C-like helper function akin to find_addr ... (and you will have to name them more explicitely) and let the stream creation to openUDPBidirectional.